### PR TITLE
Mention `as_mut` alongside `as_ref` in borrowck error message

### LINKED ISSUE #99426: Invalid suggestion to introduce `as_ref()` for `Option` used in a place expression
Consider the following code

```rust
struct Foo {
    val: i32
}

fn func1() {
    let mut foo = Some(Foo{val: 0});
    let ptr = &mut foo;
    ptr.unwrap().val += 1;
}
```

Here the diagnostic suggests to add `as_ref` as such:

```
note: this function takes ownership of the receiver `self`, which moves `*ptr`
help: consider calling `.as_ref()` to borrow the type's contents
  |
8 |     ptr.as_ref().unwrap().val += 1;
  |         +++++++++
```

This suggestion is incorrect – `as_ref` will never (AFAIK) produce a mutable reference necessary for the assignment. We should suggest `as_mut` where we know that a mutable reference will be required, instead.

### PR DESCRIPTION
**Pull request by compiler-errors**
Kinda fixes #99426 but I guess that really might be better staying open to see if we could make it suggest `as_mut` in a structured way. Not sure how to change borrowck to know that info tho.

### CODE REVIEW

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 1]
```diff
@@ -1089,7 +1089,7 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                     if is_option_or_result && maybe_reinitialized_locations_is_empty {
                         err.span_suggestion_verbose(
                             fn_call_span.shrink_to_lo(),
-                            "consider calling `.as_ref()` to borrow the type's contents",
+                            "consider calling `.as_ref()` or `.as_mut()` to borrow the type's contents",
                             "as_ref().",
                             Applicability::MachineApplicable,
```
Comment (fee1-dead) [inline compiler/rustc_borrowck/src/diagnostics/mod.rs:1094]: It looks like the user has a choice. So should we be making this `MachineApplicable`? Thinking of it again, should it be a suggestion at all or just a `span_help`?

Comment (compiler-errors) [reply compiler/rustc_borrowck/src/diagnostics/mod.rs:1094]: didn't even notice this was MachineApplicable. Yeah, it shouldn't. I could probably bump it down and make the suggestion inline, not verbose like it's rendering now

**Review by fee1-dead — APPROVED** [review]
LGTM, sorry for the late review.

**Pull request merged**